File: p105m/build_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function (preorder, inorder) {
  if (preorder.length === 0) return null;
  // find the root index at the inorder array
  let rootAtInorder = inorder.findIndex((element) => element === preorder[0]);
  let root = new Tree.TreeNode(preorder[0]);
  root.left = buildTree(
    preorder.slice(1, rootAtInorder + 1),
    inorder.slice(0, rootAtInorder)
  );
  root.right = buildTree(
    preorder.slice(rootAtInorder + 1),
    inorder.slice(rootAtInorder + 1)
  );
  return root;
};

[
  { preorder: [1], inorder: [1], expected: '1,#,#' },
  { preorder: [1, 2], inorder: [2, 1], expected: '1,2,#,#,#' },
  { preorder: [1, 2], inorder: [1, 2], expected: '1,#,2,#,#' },
  {
    preorder: [3, 9, 20, 15, 7],
    inorder: [9, 3, 15, 20, 7],
    expected: '3,9,#,#,20,15,#,#,7,#,#',
  },
].forEach(({ preorder, inorder, expected }) => {
  const actual = Tree.serialize(buildTree(preorder, inorder));
  console.log('Build tree ->', actual);
  console.assert(actual === expected);
});

File: p105m/build_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import List
from local_packages.binary_tree import TreeNode


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if not preorder:
            return None
        i = inorder.index(preorder[0])
        return TreeNode(
            val=preorder[0],
            left=self.buildTree(preorder[1 : i + 1], inorder[:i]),
            right=self.buildTree(preorder[i + 1 :], inorder[i + 1 :]),
        )


# TESTS
for preorder, inorder, expected in [
    ([1], [1], "1,#,#"),
    ([1, 2], [2, 1], "1,2,#,#,#"),
    ([1, 2], [1, 2], "1,#,2,#,#"),
    ([3, 9, 20, 15, 7], [9, 3, 15, 20, 7], "3,9,#,#,20,15,#,#,7,#,#"),
]:
    sol = Solution()
    actual = TreeNode.serialize(sol.buildTree(preorder, inorder))
    print("Build Tree ->", actual)
    assert actual == expected
